lib/plaindate.ts

```diff
@@ -225,11 +227,10 @@ export class PlainDate implements Temporal.PlainDate {
         // not a TimeZoneProtocol.
         // TODO: should we check for that shape to improve on the (bad) error
         // message that the caller will get from ToTemporalTimeZone?
-        timeZone = ES.ToTemporalTimeZone(item as Temporal.TimeZoneProtocol);
```

Comment:
General question for changes in this method: why are these changes helpful?  In general, I like minimizing changes from the original proposal-temporal source, so naively I'd prefer leaving variable names the same, but maybe I'm missing some other benefit of the changes?

Reply:
I'm assuming you are referring to the change below that uses the name `temporalTimeConverted`? This particular line just removed a cast, but the variable names should be the same as upstream.

Unfortunately, TS won't narrow the type of `temporalTime` when it sees

```ts
temporalTime = ES.ToTemporalTime(temporalTime);
```

because it can't be sure whether any future reassignments that aren't of type `PlainTime` are desired or an error.

The other way to try and keep this closer to upstream would be to add some assertion functions that will force this narrowing at the cost of adding an extra statement in some places:

```ts
assertIsTemporalPlainTime(arg: unknown): asserts arg is Temporal.PlainTime {}

...

temporalTime = ES.ToTemporalTime(temporalTime);
assertIsTemporalPlainTime(temporalTime); // This is OK because ToTemporalTime always returns a PlainTime

// temporalTime is now a PlainTime
```

This still means adding an extra statement which will mess up rebasing a bit (though other changes in this particular chunk would likely mess this up anyways).

Reply:
> I'm assuming you are referring to the change below that uses the name `temporalTimeConverted`? This particular line just removed a cast, but the variable names should be the same as upstream.

I was mostly wondering about why any of these changes were needed. What was working before this PR that's now failing with the new lint rules?

BTW, interesting idea of using an assertion function. Would it work (and if yes, would we want to do this?) to define a generic no-op assertion function that could be re-used in all cases where we need to let TS know that a variable's type has changed?

```ts
function noOpAsserter<T>(unused: unknown): asserts unused is T {};
```

Reply:
Oh, lol I see you've done something like this!

Reply:
> What was working before this PR that's now failing with the new lint rules?

I could split these PRs up, as there's two groups of changes to sources here:
 - removing casts that are doing nothing (this is what the linter is catching)
  - changing code that uses casts to avoid them and instead fix whatever the underlying problem(s) they might be masking.
  
  These changes are more in the latter bucket, where casts are being used to get the code to compile and might be hiding problems in code. Generally, I'd prefer to avoid having a bunch of casts scattered throughout the code in favour of either fixing the root cause so inference works correctly or using type-assertion functions to fix type-inference in places where there's a limitation, and documenting why that particular assertion is needed.
  
  I think the generic version would work - I'll update the PR with that.

Reply:
Makes sense. Esp. given that the no-op functions are removed in prod, I really like the assertion function thing as alternative to littering casts everywhere. Thanks for finding that!